### PR TITLE
add address region on form 'Adhesion'

### PR DESCRIPTION
**Pull request by Raulnet**
add address region on form 'Adhesion'

### CODE REVIEW

File: tests/Controller/EnMarche/MembershipControllerTest.php [hunk 1]
```diff
@@ -196,7 +196,6 @@ public function testAdherentSubscriptionTypesArePersistedCorrectlyWhenAdhesionFr
 
         $this->authenticateAsAdherent($this->client, 'simple-user@example.ch');
         $this->client->request(Request::METHOD_GET, '/adhesion');
-
```
Comment (ottaviano) [inline tests/Controller/EnMarche/MembershipControllerTest.php]: why this change?

File: src/Entity/NullablePostAddress.php [hunk 2]
```diff
@@ -112,46 +119,49 @@ public static function createFrenchAddress(
...
+            $longitude,
+            $region
         );
 
         $address->city = sprintf('%s-%s', $postalCode, $inseeCode);
 
         return $address;
     }
 
+    //TODO placer region avant lattitude est refacto l'utilisation de la method
```
Comment (ottaviano) [inline src/Entity/NullablePostAddress.php]: you can remove this

File: src/Controller/EnMarche/MembershipController.php [hunk 3]
```diff
@@ -52,7 +52,6 @@ public function registerAction(Request $request, GeoCoder $geoCoder, Authorizati
             $geoCoder->getCountryCodeFromIp($request->getClientIp()),
             $request->request->get('g-recaptcha-response')
         );
-
```
Comment (ottaviano) [inline src/Controller/EnMarche/MembershipController.php]: why this change?

File: src/Address/Address.php [hunk 4]
```diff
@@ -48,6 +48,8 @@ class Address implements AddressInterface, GeocodableInterface
      */
     private $country;
 
+    private $region;
```
Comment (ottaviano) [inline src/Address/Address.php]: `@Assert\Length(max=255)` missing?

File: front/services/address/GooglePlaceAutocomplete.js [hunk 5]
```diff
@@ -120,7 +120,9 @@ export default class GooglePlaceAutocomplete extends EventEmitter {
         } else if (!this._address.getCountry()) {
             this._address.setCountry('FR');
         }
-
+        if(this._state.country !== null && this._state.country.short_name !== 'FR') {
```
Comment (ottaviano) [inline front/services/address/GooglePlaceAutocomplete.js]: I think we can use the previous IF statement:
```js
if (this._state.country && this._state.country.short_name) {                                                  
    this._address.setCountry(this._state.country.short_name);                                                 
                                                                                                              
    if (this._state.country.short_name !== 'FR') {                                                            
        this._address.setRegion(                                                                              
            this._state.administrative_area_level_1 && this._state.administrative_area_level_1.long_name || ''
        )                                                                                                     
    }                                                                                                         
} else if (!this._address.getCountry()) {                                                                     
    this._address.setCountry('FR');                                                                           
}                                                                                                            
```

File: front/services/address/AddressObject.js [hunk 6]
```diff
@@ -1,14 +1,16 @@
 export default class AddressObject {
-    constructor(addressField, postalCodeField, cityField, countryField) {
+    constructor(addressField, postalCodeField, cityField, countryField, regionField) {
         this._address = addressField;
         this._postalCode = postalCodeField;
         this._city = cityField;
         this._country = countryField;
+        this._regionField = regionField;
```
Comment (ottaviano) [inline front/services/address/AddressObject.js]: for being consistent can you rename on `this._region`?

File: front/pages/join.js [hunk 7]
```diff
@@ -12,6 +28,48 @@ export default (formType) => {
         zipCodeField = dom('#become_adherent_address_postalCode');
     }
 
+    const autocompeleteWrapper = dom('.address-autocomplete');
+    const addressBlock = dom('.address-block');
+    const fieldCityName = dom('#adherent_registration_address_cityName');
+    const fieldRegion = dom('#adherent_registration_address_region');
```
Comment (ottaviano) [inline front/pages/join.js]: unused

File: front/pages/join.js [hunk 8]
```diff
@@ -12,6 +28,48 @@ export default (formType) => {
...
+        fieldCityName,
+        dom('#adherent_registration_address_country'),
+        dom('#adherent_registration_address_region')
+    );
+
+    if (!address.isFilled()) {
+        if ('undefined' === typeof google) {
+            return;
+        }
+        const autocomplete = new GooglePlaceAutocomplete(autocompeleteWrapper, address, 'form form--full form__field');
```
Comment (ottaviano) [inline front/pages/join.js]: I think we can refactor this part by moving this code into a new class (for ex: `AutocompletedAddressForm`) and avoid the code duplication

File: tests/Controller/EnMarche/MembershipControllerTest.php (hunk 1 above)
Comment (Raulnet) [reply tests/Controller/EnMarche/MembershipControllerTest.php]: fixed

File: src/Entity/NullablePostAddress.php (hunk 2 above)
Comment (Raulnet) [reply src/Entity/NullablePostAddress.php]: fixed

File: src/Controller/EnMarche/MembershipController.php (hunk 3 above)
Comment (Raulnet) [reply src/Controller/EnMarche/MembershipController.php]: fixed

File: src/Address/Address.php (hunk 4 above)
Comment (Raulnet) [reply src/Address/Address.php]: fixed

File: front/services/address/GooglePlaceAutocomplete.js (hunk 5 above)
Comment (Raulnet) [reply front/services/address/GooglePlaceAutocomplete.js]: fixed

File: front/services/address/AddressForm.js [hunk 9]
```diff
@@ -50,10 +62,16 @@ export default class AddressForm {
...
         if ('FR' !== this._state.country || Array.isArray(this._state.cities) && 0 === this._state.cities.length) {
             show(this._cityName);
+            if(this._region !== null) {
+                show(this._region);
+            }
             this._cityName.required = this._cityNameRequired;
             this._cityName.value = this._initialCityName;
             return;
         }
+        if(this._region !== null) {
```
Comment (ottaviano) [inline front/services/address/AddressForm.js]: CS: can you add a new empty line before, please?

File: front/services/address/AddressFormFactory.js [hunk 10]
```diff
@@ -5,14 +5,19 @@ export default class AddressFormFactory {
         this._api = api;
     }
 
-    createAddressForm(country, postalCode, city, cityName, cityNameRequired) {
+    createAddressForm(country, postalCode, city, cityName, cityNameRequired, region = null) {
+        let regionElem = null;
+        if(region !== null && dom('#'+region) !== 'undefined') {
+            regionElem = dom('#'+region);
+        }
         return new AddressForm(
```
Comment (ottaviano) [inline front/services/address/AddressFormFactory.js]: what about:
```js
createAddressForm(country, postalCode, city, cityName, cityNameRequired, region = null) {
  return new AddressForm(
    this._api,
    dom('#'+country),
    dom('#'+postalCode),
    dom('#'+city),
    dom('#'+cityName),
    cityNameRequired,
    region ? dom('#'+region) : null
  );
}
```

File: templates/javascript.js.twig [hunk 11]
```diff
@@ -5,6 +5,7 @@ App.createAddressSelector(
     '{{ address.city.vars.id }}',
     '{{ address.cityName.vars.id }}',
     '{{ address.cityName.vars.required }}',
+    '{{ address.region is defined ? address.region.vars.id : null }}',
```
Comment (ottaviano) [inline templates/javascript.js.twig]: with this, you can pass `'null'` as a string value to the construct

Comment (Raulnet) [reply templates/javascript.js.twig]: {{ address.region is defined ? address.region.vars.id|e('js') : null }},

{{ address.region is defined ? "'" address.region.vars.id|e('js') : null }},
